clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -41,6 +42,150 @@ struct is_iterator<T *> : std::true_type {};
 
 #ifndef DPCT_USM_LEVEL_NONE
 
+// device_allocator_traits is a helper struct which makes use of custom
+//  allocator constructor routines when they are specified serially on the host,
+//  while enabling oneDPL pstl accelleration when such custom constructors are
+//  not provided.
+template <typename _Allocator> struct device_allocator_traits {
+
+  // taken from libc++
+  template <class, class _Alloc, class... _Args>
+  struct __has_construct_impl : ::std::false_type {};
+
+  template <class _Alloc, class... _Args>
+  struct __has_construct_impl<decltype((void)std::declval<_Alloc>().construct(
+                                  std::declval<_Args>()...)),
+                              _Alloc, _Args...> : ::std::true_type {};
+
+  template <class _Alloc, class... _Args>
+  struct __has_construct : __has_construct_impl<void, _Alloc, _Args...> {};
+
+  template <class _Alloc, class _Pointer, class = void>
+  struct __has_destroy : ::std::false_type {};
+
+  template <class _Alloc, class _Pointer>
+  struct __has_destroy<_Alloc, _Pointer,
+                       decltype((void)std::declval<_Alloc>().destroy(
+                           std::declval<_Pointer>()))> : ::std::true_type {};
+  // end of taken from libc++
+
+  template <typename T, typename Size>
+  static
+      typename ::std::enable_if_t<__has_construct<_Allocator, T *>::value, void>
+      uninitialized_value_construct_n(_Allocator alloc, T *p, Size n) {
```

Comment:
How about implement `uninitialized_value_construct_n`, `uninitialized_fill_n`, `uninitialized_device_copy_n` inside `__has_construct` to make code cleaner and simpler?

Reply:
I'm not sure I understand this request.  How would one implement these inside of `__has_construct`?

One thing I could think that _might_ make things simpler is to convert the `::std::enable_if_t<__has_construct<_Allocator, T *>::value, void>` with different functions to `if constexpr(__has_construct<_Allocator, T *>::value)` within each of these `uniniitialized_...` functions.  This would consolidate each function to a single instance, and could consolidate currently repeated assertions.  Is this what you might be asking for?

Reply:
I did what I mentioned here above and I think I prefer it, so I kept it.  Let me know if you had something else in mind.

Reply:
LGTM